ext/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -99,222 +103,414 @@ accelerators while being able to synchronize and optimize data transfers at
+ successfully. Otherwise, it returns a NULL value with one of the
+ following error values returned in _errcode_ret_:
+
+* TODO.
 
-[quote]
-The semantics of the returned built-in kernels are undefined or defined in
-vendor-specific documentation, unless the name starts with prefix `khr_',
-which means it's a built-in kernel with semantics defined in Appendix I.
 
 === Add new appendix "Appendix I - Defined Built-in Kernels" to OpenCL API Specification
 
-This chapter describes standard built-in kernels (BiK) with well-defined
-semantics. A conformant device can report to support zero or more of the built-in
-kernels via `CL_DEVICE_BUILT_IN_KERNELS` or `CL_DEVICE_BUILT_IN_KERNELS_WITH_VERSION` device queries.
+This chapter describes standard defined built-in kernels (DBK) with
+well-defined semantics. A conformant devices can report to
+availability of the built-in kernels listed in this section with
+`clCreateDefinedBuiltInKernelDescriptor` call. The availability of a
+DBK is determined from the arguments passed to the
+`clCreateDefinedBuiltInKernelDescriptor` and unavailability of a DBK
+is indicated by CL_DBK_UNAVAILABLE error code.
 
-The general client-side abstraction of the defined built-in kernels is similar to a call
-to a C function of which implementation is hidden. The device driver can invoke one or
-more physical hardware accelerators combined with firmware to implement the semantics
-as efficiently as possible.
+The general client-side abstraction of the DBKs is similar to a call
+to a C function of which implementation is hidden. The device driver
+can invoke one or more physical hardware accelerators combined with
+firmware to implement the semantics as efficiently as possible.
 
 It is the driver's responsibility to handle efficient synchronization and communication
 to the hardware accelerator, the internal accelerator state management and resource sharing
 across multiple OpenCL contexts.
 
-==== Standard Built-in Kernels ====
+Identical DBKs with identical inputs, are not guaranteed to produce
+identical results:
+
+* across vendors,
 
-The following list of recognized built-ins is organized according to their application
-domain and handled data types. It is expected to grow and update while preserving backwards
-compatibility.
+* across driver versions and
+
+* across devices.
+
+Otherwise, identical results are produced unless:
+
+* otherwise stated in DBK's description or
+
+* the DBK has CL_DBK_PROPERTY_NON_DETERMINISTIC property set to true.
+
+Two DKBs are considered identical if their descriptors are created
+using identical kernel name, kernel attribute and kernel mode property
+arguments.
+
+==== Standard Defined Built-in Kernels ====
+
+The following list of recognized defined built-in kernels. It is
+expected to grow and update while preserving backwards compatibility.
+
+Each defined built-in kernel entry is organized as followed:
+
+* *Name*: Name of the defined built-in kernel (an enumeration).
```

Comment:
> Do we want an enumeration or is a string enough?

Enumeration is good enough and cheaper to process than C strings.

> We could expand the old BiK call and follow its usage style as close as possible?

Does not seem viable and could be very cumbersome. If we want to include NN convolution with their many attributes (1D/2D/3D mode, strides, kernel sizes dilation, padding sizes, padding modes, grouping), we would propably end up with lots of DBK kernels with awfully long names (e.g. `conv2d_kernel<x>_<y>_stride<x>_<y>_padding_<x _begin>_<x_end>_<y_begin>_<y_end>_padwith<mode>_group<grouping>...`).

Reply:
I won't consider "cheapness of processing" as a strong motivator here with today's hardware. Perhaps in the 1990s. The most important aspect is the cleanliness of the API and to try to avoid fragmentation. This will now introduce another API completely separate from the old BiK. How do the old BiKs relate to DBKs in the plan? Be a complete separate feature? We might want to think this through a bit more to avoid confusion.

Reply:
> This will now introduce another API completely separate from the old BiK. How do the old BiKs relate to DBKs in the plan? Be a complete separate feature?

What are the immediate issues having a separate API?

Reply:
A new API that is close but not exactly to the old one tends to make the interface less clean and more confusing. If we can avoid that, it's a cleaner API and can merge some functionalities with the basic BiK API, like was the original intention. In the current proposal we need another query to list the DBKs in clGetDeviceInfo(), for example, although the concepts are similar.

For example, a more advanced version of the current version with string identifier, but the configurable aspects abstracted away with the attribute mechanism (clCreateProgramWithBuiltinKernelWithProperties() ?) could work. The base name would be string based and we can add properties that restrict the choice. clGetDeviceInfo() would return the DBK name in case any attribute configuration of the DBK is supported by the device.

Reply:
I don’t mind changing the enumeration to C string after you pointed out:

> clGetDeviceInfo() would return the DBK name in case any attribute configuration of the DBK is supported by the device.

IIUC, with C string as the type of the name for DBKs, the ‘clGetDeviceInfo()’ may list the available DKBs (in some form) in `CL_DEVICE_BUILT_IN_KERNELS_WITH_VERSION` list. 

> … mechanism (clCreateProgramWithBuiltinKernelWithProperties() ?) could work. 
>
> We can always create multiple programs if we need multiple accuracy levels for some reason.

Does this mean that the `clCreateProgramWithBuiltinKernelWithProperties()` creates a single program with a single BiK or DBK in it?

It probably does, because creating a program with multiple DBK instances by the same name but by different attributes and/or properties would cause multiple definition errors. This could be avoided by assigning an unique (vendor designated) kernel name for each unique DBK and BiK that accepts attributes and properties. A `cl_kernel` handle for a specific DBK/BiK instance can be obtained without knowing the kernel name with the currently proposed `clCreateDefinedBuiltinKernel()`. (but perhaps this should be renamed to something else that also covers BiKs - e.g. `clCreateKernelWithDescriptor()`?).

Reply:
Could we have a generic `clCreateBuiltinKernelWithAttributes()' which would apply for vendor-specific BiKs (documented in vendor documentation) as well as the DBKs alive? It would create a cl_kernel from a BiK name (which can be also a DBK) and would allow setting the attributes for that cl_kernel instance.

Reply:
Does the `clCreateBuiltinKernelWithAttributes()` implicitly create a `cl_program` too? Like, could the application query the program the kernel is associated with (`CL_KERNEL_PROGRAM`)?

Reply:
I think not. I'd compare this is clCreateKernel() which creates a kernel handle which has state (the args). Here the state would also include the tuning parameters/attributes.

Reply:
I mean that we could use the (already standardized) [clCreateProgramWithBuiltinKernels](https://man.opencl.org/clCreateProgramWithBuiltInKernels.html) API for creating a program with a set of BiKs (DBKs or not) then when we instantiate kernels from that "program" (for BiKs its just a collection of BiK refs?) the cl_kernel has internal state for configuring the accelerator's different aspects. What `clBuildProgram()` does with BiKs I'd leave to the implementation internals, but would not require it to be called again.